core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -2426,4 +2421,28 @@ public String toString() {
                 ")";
         }
     }
+
+    /**
+     * FetchOffsetMetadata class is used to cache offset and its log metadata.
+     */
+    static final class OffsetMetadata {
+        private long offset;
```

Comment:
It's useful to add a comment that this offset could be different from offsetMetadata.messageOffset if it's in the middle of a batch.